master/master.go

```diff
@@ -264,36 +286,25 @@ func (ms *Master) distributeReduceTask() {
 		workerID = (workerID + 1) % nWorkers
 	}
 
-	count := 0
-LOOP:
-	for {
-		select {
-		case <-finish:
-			count += 1
-			if count == len(ms.ReduceTasks) {
-				close(finish)
-				break LOOP
+	for crashUUID := range orDone(finish, ms.crashChan, len(ms.ReduceTasks)) {
+		workers, _ = ms.availableWorkers(1)
+
+		log.Info("[Master] Re-execute Map Task from ", crashUUID, "To ", workers[0].UUID)
+		reExecuteTask, _ := taskStates.Load(crashUUID)
+		go func(task ReduceTaskInfo, id int) {
+			taskStates.Delete(crashUUID)
+			taskStates.Store(workers[id].UUID, task)
+			done := ms.client.Reduce(workers[id].IP, task.toRPC())
+			if !done {
+				task.SetState(TASK_IDLE)
+				workers[id].SetState(WORKER_UNKNOWN)
+				ms.crashChan <- workers[id].UUID
+			} else {
+				task.SetState(TASK_COMPLETED)
+				workers[id].SetState(WORKER_IDLE)
+				finish <- workers[id].UUID
 			}
-		case crashUUID := <-ms.crashChan:
-			workers, _ = ms.availableWorkers(1)
-
-			log.Info("[Master] Re-execute Map Task from ", crashUUID, "To ", workers[0].UUID)
-			reExecuteTask, _ := taskStates.Load(crashUUID)
-			go func(task ReduceTaskInfo, id int) {
-				taskStates.Delete(crashUUID)
-				taskStates.Store(workers[id].UUID, task)
-				done := Reduce(workers[id].IP, task.toRPC())
-				if !done {
-					task.SetState(TASK_IDLE)
-					workers[id].SetState(WORKER_UNKNOWN)
-					ms.crashChan <- workers[id].UUID
-				} else {
-					task.SetState(TASK_COMPLETED)
-					workers[id].SetState(WORKER_IDLE)
-					finish <- workers[id].UUID
-				}
-			}(reExecuteTask.(ReduceTaskInfo), 0)
-		}
+		}(reExecuteTask.(ReduceTaskInfo), 0)
 	}
```

Comment:
Same as Map procedure.

Reply:
I wrote tests for both cases in master_test.go and they pass you can check them too